src/PDE_BPINN.jl

```diff
@@ -0,0 +1,502 @@
+end
+
+# L2 losses loglikelihood(needed mainly for ODE parameter estimation)
+function L2LossData(Tar::PDELogTargetDensity, θ)
+    Φ = Tar.Φ
+    init_params = Tar.init_params
+    dataset = Tar.dataset
+    sumt = 0
+    L2stds = Tar.allstd[3]
+    # each dep var has a diff dataset depending on its indep var and thier domains
+    # these datasets are matrices of first col-dep var and remaining cols-all indep var
+    # Tar.init_params is needed to contruct a vector of parameters into a ComponentVector
+
+    # dataset of form Vector[matrix_x, matrix_y, matrix_z]
+    # matrix_i is of form [i,indvar1,indvar2,..] (needed in case if heterogenous domains)
+
+    # Phi is the trial solution for each NN in chain array
+    # Creating logpdf( MvNormal(Phi(t,θ),std), dataset[i] )
+    # dataset[i][:, 2:end] -> indepvar cols of a particular depvar's dataset 
+    # dataset[i][:, 1] -> depvar col of depvar's dataset
+
+    if Tar.extraparams > 0
+        if Tar.init_params isa ComponentArrays.ComponentVector
+            for i in eachindex(Φ)
+                sumt += logpdf(MvNormal(Φ[i](dataset[i][:, 2:end]',
+                            vector_to_parameters(θ[1:(end - Tar.extraparams)],
+                                init_params)[Tar.names[i]])[1,
+                            :],
+                        LinearAlgebra.Diagonal(abs2.(ones(size(dataset[i])[1]) .*
+                                                     L2stds[i]))),
+                    dataset[i][:, 1])
+            end
+            sumt
+        else
+            # Flux case needs subindexing wrt Tar.names indices(hence stored in Tar.names)
+            for i in eachindex(Φ)
+                sumt += logpdf(MvNormal(Φ[i](dataset[i][:, 2:end]',
+                            vector_to_parameters(θ[1:(end - Tar.extraparams)],
+                                init_params)[Tar.names[2][i]])[1,
+                            :],
+                        LinearAlgebra.Diagonal(abs2.(ones(size(dataset[i])[1]) .*
+                                                     L2stds[i]))),
+                    dataset[i][:, 1])
+            end
+            sumt
+        end
+    else
+        return 0
+    end
+end
+
+# priors for NN parameters + ODE constants
+function priorlogpdf(Tar::PDELogTargetDensity, θ)
+    allparams = Tar.priors
+    # Vector of ode parameters priors
+    invpriors = allparams[2:end]
+
+    # nn weights
+    nnwparams = allparams[1]
+
+    if Tar.extraparams > 0
+        invlogpdf = sum(logpdf(invpriors[length(θ) - i + 1], θ[i])
+                        for i in (length(θ) - Tar.extraparams + 1):length(θ); init = 0.0)
+
+        return (invlogpdf
+                +
+                logpdf(nnwparams, θ[1:(length(θ) - Tar.extraparams)]))
+    else
+        return logpdf(nnwparams, θ)
+    end
+end
+
+function kernelchoice(Kernel, MCMCkwargs)
+    if Kernel == HMCDA
+        δ, λ = MCMCkwargs[:δ], MCMCkwargs[:λ]
+        Kernel(δ, λ)
+    elseif Kernel == NUTS
```

Comment:
That won't match if it's a type. Needs to use `isa`

Reply:
Kernel here is the constructor name, the constructors are exported via AdvancedHMC.jl

Reply:
yes but why on the datatype? That's going to ruin inference, and doesn't allow kernel-specific parameters